components/file_system/src/rate_limiter.rs

```diff
@@ -1,133 +1,362 @@
         }
     }
+}
+
+const DEFAULT_REFILL_PERIOD: Duration = Duration::from_millis(200);
 
+#[inline]
+fn calculate_ios_per_refill(ios_per_sec: usize, refill_period: Duration) -> usize {
+    (ios_per_sec as f64 * refill_period.as_secs_f64()) as usize
+}
+
+/// Limit total IO flow below provided threshold by throttling low-priority IOs.
+/// Rate limit is disabled when total IO threshold is set to zero.
+#[derive(Debug)]
+struct RawIORateLimiter {
+    ios_through: [CachePadded<AtomicUsize>; IOPriority::VARIANT_COUNT],
+    ios_per_sec: [CachePadded<AtomicUsize>; IOPriority::VARIANT_COUNT],
+    pending_ios: [CachePadded<AtomicUsize>; IOPriority::VARIANT_COUNT],
+    protected: RwLock<RawIORateLimiterProtected>,
+}
+
+#[derive(Debug)]
+struct RawIORateLimiterProtected {
+    last_refill_time: Instant,
+}
+
+macro_rules! sleep_impl {
+    ($duration:expr, "sync") => {
+        std::thread::sleep($duration)
+    };
+    ($duration:expr, "async") => {
+        tokio::time::delay_for($duration).await
+    };
+    ($duration:expr, "non-blocking") => {
+        return 0
+    };
+}
+
+macro_rules! request_impl {
+    ($self:expr, $priority:expr, $amount:expr, $mode:tt) => {{
+        let priority_idx = $priority as usize;
+        loop {
+            let cached_ios_per_refill = $self.ios_per_sec[priority_idx].load(Ordering::Relaxed);
+            if cached_ios_per_refill == 0 {
+                return $amount;
+            }
+            let amount = std::cmp::min($amount, cached_ios_per_refill);
+            let ios_through =
+                $self.ios_through[priority_idx].fetch_add(amount, Ordering::AcqRel) + amount;
+            if ios_through <= cached_ios_per_refill {
+                return amount;
+            }
+            if let Some(locked) = $self.protected.try_read() {
+                let now = Instant::now_coarse();
+                // a small delay in case a refill slips in after `ios_per_sec` was fetched.
+                if locked.last_refill_time + Duration::from_millis(1) >= now {
+                    continue;
+                }
+                // scoped by lock, relaxed order suffice.
+                let pending =
+                    $self.pending_ios[priority_idx].fetch_add(amount, Ordering::Relaxed) + amount;
+                let since_last_refill = now.duration_since(locked.last_refill_time);
+                drop(locked);
+                if since_last_refill < DEFAULT_REFILL_PERIOD {
+                    let wait = DEFAULT_REFILL_PERIOD - since_last_refill;
+                    RATE_LIMITER_REQUEST_WAIT_DURATION
+                        .with_label_values(&[$priority.as_str()])
+                        .observe(wait.as_secs_f64());
+                    sleep_impl!(wait, $mode);
+                }
+                // our request is already recorded in `pending_ios`.
+                if pending <= cached_ios_per_refill {
+                    return amount;
+                }
```

Comment:
can we do some refactoring to move these lines out of the `if Some(locked)` block, so to avoid `drop(locked)`?